smartsim/database/lsfOrchestrator.py

```diff
@@ -0,0 +1,249 @@
+        if batch:
+            batch_settings = BsubBatchSettings(
+                nodes=db_nodes//dph, time=time, project=project
+            )
+        return batch_settings
+
+    def _build_run_settings(self, exe, exe_args, **kwargs):
+        project = kwargs.get("project", None)
+        dph = kwargs.get("dpn", 1)
+        run_args = kwargs.get("run_args", {})
+        host = kwargs.get("host", "*")
+
+        if not "nrs" in run_args.keys():
+            run_args["nrs"] = 1
+        if not "rs_per_host" in run_args.keys():
+            run_args["rs_per_host"] = 1
+        
+        run_args["tasks_per_rs"] = 1
+        bind = run_args.get("cpu_per_rs", 1)
+        if bind=="ALL_CPUS":
+            bind = "42"
+        run_args["bind"] = f"packed:{bind}"
+
+        run_args["launch_distribution"] = "packed"
+        run_settings = JsrunSettings(exe, exe_args, run_args=run_args)
+        if dph > 1:
+            # tell step to create a mpmd executable
+            run_settings.set_mpmd_args(smts_per_task=42//dph, host=host)
+        return run_settings
+
+    def _initialize_entities(self, **kwargs):
+        """Initialize DBNode instances for the orchestrator."""
+        db_nodes = kwargs.get("db_nodes", 1)
+        cluster = not bool(db_nodes < 3)
+        if int(db_nodes) == 2:
+            raise SSUnsupportedError("Orchestrator does not support clusters of size 2")
+
+        dph = kwargs.get("dpn", 1)
+        port = kwargs.get("port", 6379)
+
+        db_conf = CONFIG.redis_conf
+        exe = CONFIG.redis_exe
+        ip_module = self._get_IP_module_path()
+        ai_module = self._get_AI_module()
+
+        for db_id in range(db_nodes//dph):
+            db_node_name = "_".join((self.name, str(db_id)))
+            # create the exe_args list for launching multiple databases
+            # per node. also collect port range for dbnode
+            if self.force_port_increment:
+                base_port = int(port) + db_id*dph
+            else:
+                base_port = int(port)
+            ports = []
+            exe_args = []
+            for port_offset in range(dph):
+                next_port = base_port + port_offset
+                node_exe_args = [
+                    db_conf,
+                    ai_module,
+                    ip_module,
+                    "--port",
+                    str(next_port),
+                ]
+                if cluster:
+                    node_exe_args += self._get_cluster_args(db_node_name, next_port)
+                exe_args.append(node_exe_args)
+                ports.append(next_port)
+
+            if dph == 1:
+                exe_args = exe_args[0]
+
+            # host=db_id+1 because 0 is login node
+            run_settings = self._build_run_settings(exe, exe_args, host=db_id+1, **kwargs)
+            node = DBNode(db_node_name, self.path, run_settings, ports)
+            self.entities.append(node)
+        self.ports = ports
+
+    def __len__(self):
```

Comment:
OK, I was doing this because when we use ERF files (like SPMD or MPMD), then each DBNode respresents many shards and I wanted `__len__` to represent the number of shards: it is used when checks are performed (e.g. in create_cluster). Either we add a property like `num_shards` to all Orchestrators, or I need to keep it like this, for now.

Reply:
Property might be best I think. Could we implement this in the base ``Orchestrator`` class?

```python
@property
def num_shards(self):
```    

I am worried about overriding ``__len__`` and writing against the expected ``EntityList`` interface.

Reply:
Yes, makes sense. I'll add this in the base class.